### PR TITLE
[WIP] Experimental prerenderer

### LINKED ISSUE #1482: RFC: Experimental Fast renderer
**Do you want to request a _feature_ or report a _bug_?**

RFC

**What is the current behaviour?**

Currently pre-rendering is handled by `HTMLWebpackPlugin` and `webpack`. This is quite memory expensive and works on main thread. On a few random tests the `preact build` command simply crashes when asked to create a build with 500-1000 pages and some sizable amount of pre-render data.

Even in the scenario where it does not crash it takes quite a long time to finish the build, sometimes even 30 minutes or more. See #1463 .

**What is the expected behaviour?**

Pre-rendering should not be this memory expansive and should finish much much early on. See gatsby/next and others for inspiration.

We can probably move pre-rendering out of the webpack toolchain and use nodejs worker threads to pre-render the pages in parallel. The new rendering will have a set API for the `ejs` templates which can be guarded with SEMVER. All pages can pre-render in their own worker threads and the files can be written at the end.

We can start with an experimental option in V3 and land it as stable in V4.

e.g. 
`preact build --experimental-fast-render`.

The API for the `ejs` templates would be as follows.

```js
{
  options: {
    webpack: {
      public_path: '...',
      assets: [...], //list of all assets generated by webpack
    },
    data: { 
      url: '...',
      title: '...'.
      CLI_DATA: {}, // all the pre-render data for this route
    }
  }
}
```

We would also morph this data to slightly match the `HTMLWebpackPlugin`'s existing schema so that if anybody is trying to use this option without much change in `template.html` then they would be able to.

### PR DESCRIPTION
**Pull request by prateekbh**
Fixes #1482 

**What kind of change does this PR introduce?**

Feature request.

**Did you add tests for your changes?**

Yes

**Summary**

Implements an experimental pre-renderer with a set of API for the data.

**Does this PR introduce a breaking change?**

No

**Other information**

There is some effort to keep the current templates non-breaking but the API will eventually need to be updated and can onlhy land as stable in v4.0

**API**
The `template.html` will have access to the following API
```js
{
  options: {
    url,
    manifest,
    ssr: () => {},
    CLI_DATA: { url, ...routeData },
    webpack: {publicPath: '', asset: ['all generated assets']},
    
  }
}
```

### CODE REVIEW

File: packages/cli/lib/lib/fast-render/pool.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const { cpus } = require('os');
```
Comment (prateekbh) [inline packages/cli/lib/lib/fast-render/pool.js:1]: put copyright here

File: packages/cli/lib/lib/fast-render/renderer.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+				CLI_DATA: {
+					preRenderData: data,
+				},
+			};
+			return prerender({ cwd, dest, src }, params);
+		},
+		CLI_DATA: { url, ...routeData },
+		webpack,
+	};
+	const htmlWebpackPlugin = {
```
Comment (prateekbh) [inline packages/cli/lib/lib/fast-render/renderer.js:35]: explain this

**Comment by psabharwal123** [comment]
@prateekbh can we also upgrade to html-webpack-plugin 4? https://dev.to/jantimon/html-webpack-plugin-4-has-been-released-125d

who can help review this change? would really like to try it out, our build with 100 pages takes a long time currently

**Comment by psabharwal123** [comment]
@prateekbh why was this closed? What is the solution for builds that take a very long time?

**Comment by rschristian** [comment]
@psabharwal123 Your best bet would be to move over to [WMR](https://github.com/preactjs/wmr) likely. We don't have infinite resources to develop everything in parallel, and WMR has a lot more momentum as well as advantages.

**Comment by psabharwal123** [comment]
Is there a guide to migrating from preact cli to WMR? Where can I see the advantages of it over preact-cli so I can convince our team to look into this?

**Comment by rschristian** [comment]
There is no guide at the moment, nor is there a list of advantages. WMR wasn't necessarily meant to be a successor to Preact-CLI, though that's now where more effort is being spent these days. Because of that no one's really taken time to try to sell WMR as an upgraded CLI, though maybe that's something that should be done. I'll see how people feel about that.

The upgrade might be quite a bit of work, as WMR is not Webpack-based, but Rollup-based. Any plugins you've added in your `preact.config.js` will need to find Rollup equivalents, or you could of course write your own.

**Comment by LionsAd** [comment]
@psabharwal123 

Unfortunately the experimental renderer never did make it in, but if someone still needs a fast prerender.js script:

https://github.com/preactjs/preact-cli/issues/1684#issuecomment-1094349753

**Comment by psabharwal123** [comment]
Thanks! I ll look into it

**Pull request closed without merging**